### PR TITLE
bpo-42212: check if generated files are up-to-date in github actions

### PR DESCRIPTION
**Pull request by FFY00**
https://github.com/python/core-workflow/issues/380

https://bugs.python.org/issue42212

### CODE REVIEW

**Comment by vstinner** [comment]
> Tests / Windows (x64) (pull_request) Failing after 1m — Windows (x64) 

Network issue. I re-run the job.

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -47,6 +47,35 @@ jobs:
...
+      - uses: actions/checkout@v2
+      - uses: actions/setup-python@v2
+      - name: Register gcc problem matcher
+        run: echo "::add-matcher::.github/problem-matchers/gcc.json"
+      - name: Install Dependencies
+        run: sudo ./.github/workflows/posix-deps-apt.sh
+      - name: Build CPython
+        run: |
+          # -Og is much faster than -O
+          CFLAGS="${CFLAGS} -Og" ./configure --with-pydebug
```
Comment (vstinner) [inline .github/workflows/build.yml]: Python uses -Og by default when pydebug is used, you can remove it:

```suggestion
          ./configure --with-pydebug
```

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -47,6 +47,35 @@ jobs:
...
+      - name: Check for changes
+        run: |
+          changes=$(git status --porcelain)
+          # Check for changes in regenerated files
+          if ! test -z "$changes"
+          then
+            echo "Generated files not up to date"
+            echo "$changes"
+            exit 1
+          fi
```
Comment (vstinner) [inline .github/workflows/build.yml:74]: Please check also for "make smelly" in another step. It's not directly related, but IMO this job is a good place for such check.

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -47,6 +47,35 @@ jobs:
...
+  check_generated_files:
+    name: 'Check if generated files are up to date'
+    runs-on: ubuntu-latest
+    needs: check_source
+    if: needs.check_source.outputs.run_tests == 'true'
+    steps:
+      - uses: actions/checkout@v2
+      - uses: actions/setup-python@v2
+      - name: Register gcc problem matcher
+        run: echo "::add-matcher::.github/problem-matchers/gcc.json"
```
Comment (vstinner) [inline .github/workflows/build.yml]: I'm not sure that it's a good idea to get GCC warnings here. Other jobs already produce enough warnings (and the same warnings), no?

File: .github/workflows/build.yml (hunk 1 above)
Comment (FFY00) [reply .github/workflows/build.yml]: ack :blush:

File: .github/workflows/build.yml (hunk 3 above)
Comment (FFY00) [reply .github/workflows/build.yml]: Yeah, I copied this over from the build job. I was not entirely sure what this step did, but as I was having some issues I just copied everything.

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -47,6 +47,34 @@ jobs:
...
+          ./configure --with-pydebug
+          make -j4 regen-all
+      - name: Check for changes
+        run: |
+          changes=$(git status --porcelain)
+          # Check for changes in regenerated files
+          if ! test -z "$changes"
+          then
+            echo "Generated files not up to date"
+            echo "$changes"
```
Comment (vstinner) [inline .github/workflows/build.yml:72]: Can you add a sentence to suggest running "make regen-all"?

Comment (FFY00) [reply .github/workflows/build.yml:72]: Whops, sorry for the delay, this fell off my working pool.

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -47,6 +47,34 @@ jobs:
...
+          changes=$(git status --porcelain)
+          # Check for changes in regenerated files
+          if ! test -z "$changes"
+          then
+            echo "Generated files not up to date. Perhaps you forgot to run make regen-all ;)"
+            echo "$changes"
+            exit 1
+          fi
+      - name: Check exported libpython symbols
+        run: make smelly
```
Comment (vstinner) [inline .github/workflows/build.yml:76]: I wanted to suggest to add a message explaining that the check ensures that libpython only exports starting with Py and _Py. But instead, I wrote PR #23423 to enhance the script directly ;-)

**Comment by vstinner** [comment]
Merged, thanks!

**Pull request merged**